lte/cloud/go/services/nprobe/collector/collector.go

```diff
@@ -0,0 +1,156 @@
+/*
+Copyright 2020 The Magma Authors.
+
+This source code is licensed under the BSD-style license found in the
+LICENSE file in the root directory of this source tree.
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package collector
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"magma/lte/cloud/go/services/nprobe"
+	"magma/orc8r/cloud/go/orc8r"
+	"magma/orc8r/cloud/go/services/eventd/obsidian/handlers"
+	"magma/orc8r/cloud/go/services/eventd/obsidian/models"
+	orc8r_config "magma/orc8r/lib/go/service/config"
+
+	"github.com/golang/glog"
+	"github.com/olivere/elastic/v7"
+	"github.com/thoas/go-funk"
+)
+
+// multi-stream endpoint query args
+const (
+	defaultQuerySize = 50
+
+	sortTag             = "@timestamp"
+	pathParamStreamName = "stream_name"
+	queryParamEventType = "event_type"
+
+	// We use the ES "keyword" type for exact match
+	dotKeyword              = ".keyword"
+	elasticFilterStreamName = pathParamStreamName + dotKeyword
+	elasticFilterEventType  = queryParamEventType + dotKeyword
+	elasticFilterEventTag   = "event_tag" + dotKeyword // We use event_tag as fluentd uses the "tag" field
+	elasticFilterTimestamp  = "@timestamp"
+)
+
+// multi-stream query parameters
+type multiStreamEventQueryParams struct {
+	networkID string
+	streams   []string
+	events    []string
+	tags      []string
+	timestamp int64
+}
+
+func getElasticClient() (*elastic.Client, error) {
+	// Retrieve elastic config
+	elasticConfig, err := orc8r_config.GetServiceConfig(orc8r.ModuleName, "elastic")
+	if err != nil {
+		return nil, fmt.Errorf("Failed to instantiate elastic config")
+	}
+	elasticHost := elasticConfig.MustGetString("elasticHost")
+	elasticPort := elasticConfig.MustGetInt("elasticPort")
+
+	client, err := elastic.NewSimpleClient(elastic.SetURL(fmt.Sprintf("http://%s:%d", elasticHost, elasticPort)))
+	if err != nil {
+		return nil, fmt.Errorf("Failed to instantiate elastic client")
+	}
+	return client, nil
+}
+
+// getMultiStreamsQueryParameters takes a networkID, timestamp and list of tags
+// and return a multi streams query parameters
+func getMultiStreamsQueryParameters(networkID string, timestamp int64, tags []string) multiStreamEventQueryParams {
+	return multiStreamEventQueryParams{
+		networkID: networkID,
+		streams:   nprobe.GetESStreams(),
+		events:    nprobe.GetESEventTypes(),
+		tags:      tags,
+		timestamp: timestamp,
+	}
+}
+
+// EventsCollector wraps an elastic client, query multiple streams per
+// subscriber and retrieve all events since last timestamp
+type EventsCollector struct {
+	ElasticClient *elastic.Client
+}
```

Comment:
This functionality should sit in `eventd`, and be exposed through a gRPC interface

Reply:
I think my main concern here is that we're duplicating a decent amount of code from eventd. I think we should look at more effectively exposing a client and the necessary struct types into a package (`e.g. eventd_client`) in eventd and then import those so that we can avoid this duplication.

I'm not sure that it's worthwhile to have both HTTP and gRPC implementations in eventd that wrap elasticsearch querying. As a similar example, if an orchestrator service (i.e. `analytics`) needs to query metrics, it makes more sense for that service just to use a a prometheus client, rather than adding a gRPC servicer to metricsd which would wrap the prometheus client. I feel similarly here.

Reply:
Okay, let's go with that